hw/xbox/dsp/dsp_dma.c

```diff
@@ -241,13 +242,35 @@ static void dsp_dma_run(DSPDMAState *s)
                 }
             }
 
-            // write to scratch memory
-            s->scratch_rw(s->scratch_rw_opaque,
-                scratch_buf, scratch_addr, count*item_size, 1);
+            /* FIXME: Move to function; then reuse for both directions */
+            switch (buf_id) {
+            case 0x0:
+            case 0x1:
+            case 0x2:
+            case 0x3: {
+                unsigned int fifo_index = buf_id;
+                s->fifo_rw(s->rw_opaque,
+                    scratch_buf, fifo_index, transfer_size, 1);
```

Comment:
nit: Some comment w.r.t. named constants and value for direction field. Helps with readability.

Reply:
I agree, but this is a problem across the entire codebase.
Also affects nv2a for example.

We should fix this everywhere in a larger refactor.

Reply:
Why not just use the constants we already have here?

```
    DMA_DIRECTION_TO_DEVICE = 0,
    DMA_DIRECTION_FROM_DEVICE = 1,
```

Reply:
Because I wasn't aware of those when I wrote the code.
I was only made aware of the existing QEMU DMA code by espes after this PR was finished. We don't use it in nv2a either.

Note that my code is just a variation of the existing scratch read/write code. So the same signature and code-style was kept.
So the other argument still applies: We should just refactor this in a larger refactor across our entire codebase.

However, for the DSP side, we'd also have to review the terminology in the DSP world before using such constants: the names could be confusing as the DSP might also call the host CPU side "device space" or "peripheral space", as the DSP is often equivalent to the main CPU.
For nv2a it can be equally confusing because some of the DMA is not `FROM_DEVICE` or `TO_DEVICE` but `FROM_CACHE` and `TO_CACHE` etc.

espes also recommended to use the DMA functions internally in the APU DMA code. That seems like a good idea, but needs testing, because this could be performance critical and might need some specific behavior, too.